ega-data-api-dataedge/src/main/java/eu/elixir/ega/ebi/dataedge/service/internal/EBINuFileService.java

```diff
@@ -0,0 +1,199 @@
+            throw new FileNotAvailableException(fileId);
+
+        return file;
+    }
+
+    private long getPlainFileSize(File file, String encryptionFormat) throws UnretrievableFileException {
+        long size = file.getFileSize();
+
+        if (ENCRYPTION_AES128.equals(encryptionFormat) || ENCRYPTION_AES256.equals(encryptionFormat)) {
+            size -= 16L;
+        } else if (!ENCRYPTION_PLAIN.equals(encryptionFormat)) {
+            throw new UnretrievableFileException(file.getFileId());
+        }
+
+        return size;
+    }
+
+    @Override
+    public InputStream getSpecificByteRange(String fileId, long start, long end) throws EgaFileNotFoundException, UnretrievableFileException, FileNotAvailableException, RangesNotSatisfiableException {
+        File file = getFile(fileId);
+        String encryptionAlgorithm = keyService.getEncryptionAlgorithm(fileId);
+
+        long plainFileSize = getPlainFileSize(file, encryptionAlgorithm);
+        if (start < 0 || end >= plainFileSize || end < start)
+            throw new RangesNotSatisfiableException(fileId, start, end);
+
+        try {
+            switch (encryptionAlgorithm) {
+                case ENCRYPTION_PLAIN: {
+                    return fireClientService.downloadByteRangeByPath(file.getDisplayFilePath(), start, end);
+                }
+                case ENCRYPTION_AES128:
+                case ENCRYPTION_AES256: {
+                    return decryptAES(file, encryptionAlgorithm, start, end);
+                }
+                default:
+                    throw new UnretrievableFileException(file.getFileId());
+            }
+        } catch (FileNotFoundException e) {
+            throw new EgaFileNotFoundException(fileId, e);
+        } catch (IOException | FireServiceException | URISyntaxException | ClientProtocolException e) {
+            throw new FileNotAvailableException(fileId, e);
+        }
+    }
+
+    private InputStream decryptAES(File file, String algorithm, final long start, final long end) throws IOException, FileNotAvailableException, URISyntaxException, FireServiceException, ClientProtocolException {
+
+        // AES works in blocks of 16 bytes so we have to request always whole blocks
+        final int startPadding = (int) (start % 16);
+        final int endPadding = (int) (16 - (end % 16) - 1);
+        final long fireStartByte = start - startPadding;
+        final long fireEndByte = end + endPadding;
+
+        // Add 16 because the first 16 bytes are the IV header
+        InputStream encryptedStream = fireClientService.downloadByteRangeByPath(file.getDisplayFilePath(),
+                fireStartByte + 16,
+                fireEndByte + 16);
+
+        SecretKey key = Glue.getInstance().getKey(keyService.getFileKey(file.getFileId()).toCharArray(),
+                ENCRYPTION_AES128.equals(algorithm) ? 128 : 256);
+
+        byte[] header = aesHeaderCache.get(file.getDisplayFilePath());
+        assert header != null;
```

Comment:
The header cache can be returned null if the cache had a failure when retrieving the information. This means that the cache won't try to refresh this value during 24 hours without anyone reading this value. This is known as cache poisoning as the cache will consider this state as correct and our code will keep failing at the assert on line 151.

If the value returned by the cache is null, we have to evict it manually
`cache.invalidate(key)`

After invalidation I would suggest that instead of an assert assertion we throw an exception that can be captured and understood that needs to be represented as a 503 error later, as this will only happen due to problems to read information from fire. Maybe we can use the FireServiceException